vpr/src/route/route_tree_timing.cpp

```diff
@@ -241,7 +250,12 @@ t_rt_node* update_route_tree(t_heap* hptr, SpatialRouteTreeLookup* spatial_rt_lo
 
 void add_route_tree_to_rr_node_lookup(t_rt_node* node) {
     if (node) {
-        VTR_ASSERT(rr_node_to_rt_node[node->inode] == nullptr || rr_node_to_rt_node[node->inode] == node);
+        auto& device_ctx = g_vpr_ctx.device();
+        if (device_ctx.rr_nodes[node->inode].type() == SINK) {
```

Comment:
Need to explain the logic of this ASSERT.  The rr_node that is used in this rt_node has either (1) not been added into the routing (rt_tree) for this net before, or if it was added the rr_node_to_rt_node mapping structure will point back at the rt_node we're processing so we're just branching off that point.  Exception: for SINK nodes, some netlists and input pin equivalences can lead to us routing to the same SINK more than once on a net.  Hence for SINKs we assert on a weaker condition that if this SINK is already in the rt_tree, the rr_node_to_rt_node mapping structure points to a legal rt_node (but not necessarily the only one) containing the SINK.

Reply:
Now that I think about it, this function serves to record ALL nodes from the rt_tree (built from traceback) in the rr_node_to_rt_node lookup, this is to get the mapping ready so that new branches can be added knowing where the branching points are in the existing tree. This is done recursively in a depth-first traversal, so technically, wouldn't it be illegal to reach the exact same rt_node twice no matter what? I think that the assertion should actually be changed to just

```
VTR_ASSERT( rr_node_to_rt_node[node->inode] == nullptr ||
        (rr_node_to_rt_node[node->inode] != node && rr_node_to_rt_node[node->inode]->inode == node->inode) );

```

Reply:
Or something like
```
if (device_ctx.rr_nodes[node->inode].type() == SINK) {
   VTR_ASSERT( rr_node_to_rt_node[node->inode] == nullptr ||
        (rr_node_to_rt_node[node->inode] != node && rr_node_to_rt_node[node->inode]->inode == node->inode) );
} else {
    VTR_ASSERT(rr_node_to_rt_node[node->inode] == nullptr);
}
```
so that we make sure to check it's a SINK node to use the weak assert condition.